wicket/src/runner.rs

```diff
@@ -0,0 +1,253 @@
+    // services.
+    state: State,
+
+    // A mechanism for interacting with `wicketd`
+    #[allow(unused)]
+    wicketd: WicketdHandle,
+
+    // When [`Runner::run`] is called, this is extracted and moved into a tokio
+    // task.
+    wicketd_manager: Option<WicketdManager>,
+
+    // The terminal we are rendering to
+    terminal: Term,
+
+    // Our friendly neighborhood logger
+    log: slog::Logger,
+
+    // The tokio runtime for everything outside the main thread
+    tokio_rt: tokio::runtime::Runtime,
+}
+
+#[allow(clippy::new_without_default)]
+impl Runner {
+    pub fn new(log: slog::Logger, wicketd_addr: SocketAddrV6) -> Runner {
+        let (events_tx, events_rx) = channel();
+        let state = State::new();
+        let backend = CrosstermBackend::new(stdout());
+        let terminal = Terminal::new(backend).unwrap();
+        let tokio_rt = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+        let (wicketd, wicketd_manager) =
+            WicketdManager::new(&log, events_tx.clone(), wicketd_addr);
+        Runner {
+            screen: Screen::new(),
+            events_rx,
+            events_tx,
+            state,
+            wicketd,
+            wicketd_manager: Some(wicketd_manager),
+            terminal,
+            log,
+            tokio_rt,
+        }
+    }
+
+    pub fn run(&mut self) -> anyhow::Result<()> {
+        self.start_tokio_runtime();
+        enable_raw_mode()?;
+        execute!(self.terminal.backend_mut(), EnterAlternateScreen,)?;
+        self.mainloop()?;
+        disable_raw_mode()?;
+        execute!(self.terminal.backend_mut(), LeaveAlternateScreen,)?;
+        Ok(())
+    }
+
+    fn mainloop(&mut self) -> anyhow::Result<()> {
+        info!(self.log, "Starting main loop");
+
+        // Draw the initial screen
+        self.screen.draw(&self.state, &mut self.terminal)?;
+
+        loop {
+            // unwrap is safe because we always hold onto a Sender
+            let event = self.events_rx.recv().unwrap();
+            match event {
+                Event::Tick => {
+                    let action = self.screen.on(&mut self.state, Event::Tick);
+                    self.handle_action(action)?;
+                }
+                Event::Term(TermEvent::Key(key_event)) => {
+                    if is_control_c(&key_event) {
+                        info!(self.log, "CTRL-C Pressed. Exiting.");
+                        break;
+                    }
+                    let action = self.screen.on(
+                        &mut self.state,
+                        Event::Term(TermEvent::Key(key_event)),
+                    );
+                    self.handle_action(action)?;
+                }
+                Event::Term(TermEvent::Resize(_, _)) => {
+                    self.screen.draw(&self.state, &mut self.terminal)?;
+                }
+                Event::Inventory(event) => {
```

Comment:
Tiny style nit - this match arm is arguably long enough (and has enough rightward-drift) it warrants its own method.

Reply:
Fixed in [fd4ea2d](https://github.com/oxidecomputer/omicron/pull/2440/commits/fd4ea2d1c6e5f6a298b987e48eeefb5ee6bbeaf0)